xds/internal/xdsclient/updates.go

```diff
@@ -0,0 +1,504 @@
+func (l ListenerUpdate) Equal(other ListenerUpdate) bool {
+	switch {
+	case l.RouteConfigName != other.RouteConfigName:
+		return false
+	case !l.InlineRouteConfig.Equal(other.InlineRouteConfig):
+		return false
+	case l.MaxStreamDuration != other.MaxStreamDuration:
+		return false
+	case !l.InboundListenerCfg.Equal(other.InboundListenerCfg):
+		return false
+	}
+	if len(l.HTTPFilters) != len(other.HTTPFilters) {
+		return false
+	}
+	for i := 0; i < len(l.HTTPFilters); i++ {
+		if !l.HTTPFilters[i].Equal(other.HTTPFilters[i]) {
+			return false
+		}
+	}
+	return true
+}
+
+// HTTPFilter represents one HTTP filter from an LDS response's HTTP connection
+// manager field.
+type HTTPFilter struct {
+	// Name is an arbitrary name of the filter.  Used for applying override
+	// settings in virtual host / route / weighted cluster configuration (not
+	// yet supported).
+	Name string
+	// Filter is the HTTP filter found in the registry for the config type.
+	Filter httpfilter.Filter
+	// Config contains the filter's configuration
+	Config httpfilter.FilterConfig
+}
+
+// Equal returns true if h and other are the same.
+func (h HTTPFilter) Equal(other HTTPFilter) bool {
+	switch {
+	case h.Name != other.Name:
+		return false
+	case !grpcutil.EqualAny(h.Filter, other.Filter, func(x, y interface{}) bool {
+		return grpcutil.EqualStringSlice(x.(httpfilter.Filter).TypeURLs(), y.(httpfilter.Filter).TypeURLs())
+	}):
+		return false
+	case !grpcutil.EqualAny(h.Config, other.Config, func(x, y interface{}) bool {
+		return x.(httpfilter.FilterConfig).Equal(y.(httpfilter.FilterConfig))
+	}):
+		return false
+	}
+	return true
+}
+
+// InboundListenerConfig contains information about the inbound listener, i.e
+// the server-side listener.
+type InboundListenerConfig struct {
+	// Address is the local address on which the inbound listener is expected to
+	// accept incoming connections.
+	Address string
+	// Port is the local port on which the inbound listener is expected to
+	// accept incoming connections.
+	Port string
+	// FilterChains is the list of filter chains associated with this listener.
+	FilterChains *FilterChainManager
+}
+
+// Equal returns true if i and other are the same.
+func (i *InboundListenerConfig) Equal(other *InboundListenerConfig) bool {
+	if i == other {
+		return true
+	}
+	if i == nil || other == nil {
+		return false
+	}
+	switch {
+	case i.Address != other.Address:
+		return false
+	case i.Port != other.Port:
+		return false
+	case !i.FilterChains.Equal(other.FilterChains):
+		return false
+	}
+	return true
+}
+
+// RouteConfigUpdate contains information received in an RDS response, which is
+// of interest to the registered RDS watcher.
+type RouteConfigUpdate struct {
+	VirtualHosts []*VirtualHost
+	// Raw is the resource from the xds response.
+	Raw *anypb.Any
+}
+
+// Equal returns true if r and other are the same.
+func (r *RouteConfigUpdate) Equal(other *RouteConfigUpdate) bool {
```

Comment:
How did you decide whether it should be pointer receiver or not?
E.g. `ListenerUpdate` is not pointer receiver, but this is.

Reply:
I decided based on how each of these types are stored within other types. `ListenerUpdate` is a top-level type which is not stored as a pointer inside any other type, while `RouteConfigUpdate` is stored as a pointer inside `ListenerUpdate`.